### PR TITLE
Add missing algorithm history tracking to UnGroupWorkspace

### LINKED ISSUE #34255: History for UnGroupWorkspace gets lost
Create a group workspace
Run UnGroupWorkspace on it
The resulting workspace(s) don't show the run of UnGroupWorkspace in their history

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws1 = CreateSampleWorkspace()
ws2 = CreateSampleWorkspace()
group_ws = GroupWorkspaces(InputWorkspaces="ws1,ws2")
UnGroupWorkspace(group_ws)
```

I noticed this when running the Indirect Data Reduction UI (Energy Transfer tab) but the problem is more generally related to the UnGroupWorkspace algorithm I think. If you try and script the history of a workspace produced by this tab you get a group workspace whereas the UI gives a single ungrouped workspace because it does an additional call to UnGroupWorkspace that's missing from the history.

### PR DESCRIPTION
**Pull request by jhaigh0**
**Description of work**

Fixes a bug where workspaces which belonged to a group that was passed to `UnGroupWorkspace`, did not include the algorithm in their history.

The reason why is that when passing on algorithm history, the Algorithm base class searches for algorithm properties which are outputs and can cast to a workspace. `UnGroupWorkspace` simply removes the group workspace passed to it (releasing the child workspaces in doing so) so doesn't need an 'output' property, meaning the history is never passed on.

**Summary of work**
The simplest way has involved moving the second `fillHistory` declaration from private to protected. I don't see this as being dangerous to do anyway, just disrupts the 'correctness' a little.
Also added in a test.

**To test:**

Run this script
```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws1 = CreateSampleWorkspace()
ws2 = CreateSampleWorkspace()
group_ws = GroupWorkspaces(InputWorkspaces="ws1,ws2")
UnGroupWorkspace(group_ws)
```
Right click `ws1` and click `Show History`; `UnGroupWorkspace` should be the last algorithm on there with correct input worksapce listed.
Check again for `ws2`.

Fixes #34255

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Algorithms/test/UnGroupWorkspaceTest.h [hunk 1]
```diff
@@ -125,4 +148,16 @@ class UnGroupWorkspaceTest : public CxxTest::TestSuite {
...
+  bool isAlgorithmInHistory(const Mantid::API::Workspace_sptr ws) {
+    auto ws_histories = ws->getHistory().getAlgorithmHistories();
+    bool found = false;
+    for (auto alg_history : ws_histories) {
+      if (alg_history->name() == "UnGroupWorkspace") {
+        found = true;
+        break;
+      }
+    }
+    return found;
```
Comment (robertapplin) [inline Framework/Algorithms/test/UnGroupWorkspaceTest.h:161]: I think we can use a c++ std library algorithm here
```suggestion
    auto iter = std::find_if (wsHistories.cbegin(), wsHistories.cend(), [](auto const &algHistory){ return algHistory->name() == "UnGroupWorkspace";});
    return iter != wsHistories.cend();
```

File: Framework/Algorithms/test/UnGroupWorkspaceTest.h [hunk 2]
```diff
@@ -125,4 +148,16 @@ class UnGroupWorkspaceTest : public CxxTest::TestSuite {
         ads.remove(member);
     }
   }
+
+  bool isAlgorithmInHistory(const Mantid::API::Workspace_sptr ws) {
```
Comment (robertapplin) [inline Framework/Algorithms/test/UnGroupWorkspaceTest.h:152]:
```suggestion
  bool isAlgorithmInHistory(const Mantid::API::Workspace_sptr &ws) {
```

**Review by robertapplin — CHANGES_REQUESTED** [review]
Nice find and thanks for adding a test. I have a few suggestions but looks good overall

File: Framework/Algorithms/src/UnGroupWorkspace.cpp [hunk 3]
```diff
@@ -53,6 +52,11 @@ void UnGroupWorkspace::exec() {
   if (!wsGrpSptr) {
     throw std::runtime_error("Selected Workspace is not a WorkspaceGroup");
   }
+  // add algorithm history to the workspaces being released from the group
+  WorkspaceVector outputWorkspaces;
+  for (const auto &ws : wsGrpSptr->getAllItems())
+    outputWorkspaces.emplace_back(ws);
+  fillHistory(outputWorkspaces);
```
Comment (robertapplin) [inline Framework/Algorithms/src/UnGroupWorkspace.cpp:59]: Is it possible to just do this?
```suggestion
  fillHistory(wsGrpSptr->getAllItems());
```

Comment (jhaigh0) [reply Framework/Algorithms/src/UnGroupWorkspace.cpp:59]: yes 🙃 clearly did that when tired

**Review by robertapplin — APPROVED** [review]
Thanks for the changes, happy to approve

**Pull request merged**